### ISSUE TITLE
Exchange private data inside maps with gibberish data -> for sharing buggy maps without publishing private data

### ISSUE THREAD

**Issue by autoca** [freeplane/freeplane#702]
Hi Dimitry,

This FR (feature request) is about simplifiying sharing maps in case of perceived bugs or for similar reasons.

**Is your feature request related to a problem? Please describe.**
When something in a mindmap does not behave as expected, I might want to share that map with the community and, especially, with you, Dimitry, here in the forums, so that you or others can more easily understand the problematic behaviour (possible bug). However, mostly, this happens with maps that have private information in them (either work related or personal). It is cumbersome to have to "rebuild" such a map (even if just rebuilding part of it) in order to not have private data in them.

**Describe the solution you'd like**
It would be nice if a feature existed that would allow me to clone a map and at the same time exchange any existing information in that map with gibberish text and maybe "default" images. This of course, should not be more work for the user than one click on the corresponding menu item.

**Describe alternatives you've considered**
Doing it all by hand. This is cumbersome.

**Additional context**
None. I hope it is clear enough.

MTIA and regards, a.

**Comment by macmarrum** [comment]
I think this can be done with a script. I'll give it a try.

**Comment by macmarrum** [comment]
@autoca Here's a script that creates a copy of the current mindmap with the name prefix "obfuscated~"
- https://github.com/macmarrum/freeplane-scripting/blob/main/src/scripts/obfuscateMindmapForSharing.groovy

Please test it and provide your feedback.

**Comment by autoca** [comment]
This is awesome effort. Thank you very much, macmarrum!
Unfortunaetly, I again have a permission issue that, as I mentioned in the other thread, do not really want to ignore as long as I would have to do it generally for all scripts:
![grafik](https://user-images.githubusercontent.com/37442404/193362036-a97ec049-2368-4173-b72b-f3dcf8da69c9.png)

**Comment by macmarrum** [comment]
A script can be assigned access permissions only if it's part of an add-on. Otherwise the assumption is that User trusts the scripts User runs and the permissions are granted globally.

I've made the script into an add-on -> https://github.com/macmarrum/freeplane-MapObfuscator
Please let me know how it works for you.

**Comment by autoca** [comment]
Wow, thank you very much!

I did tested it but not very thoroughly yet. On first glance, it seems to work nicely, but I found 3 things that I would improve if I was the author:

1) The root node does not really get obfuscated, since adding a prefix is not obfuscating what comes after it. But, of course, it is not a big issue, since that one node can be done by hand.

2) Images are lost. Of course the original pics should probably be changed, but it would be good if a default test img would be inserted instead of each original. Sometimes bugs or unexpected behaviors just have to do with images and it cannot be tested without any images present. Since this probably does not happen too often and if so, probably not with many images, it probably can be done by hand as well.

3) When the add-on is executed, it says that a map with the new name already exists and asks if it should be overwritten. Of course, there is no such map already in existance. A simple confirmation with "Yes" is enough for it to work, but it is a bit confusing.

Again, great work and big thank you!
Regards, a.

**Comment by macmarrum** [comment]
Thanks for your feedback. It's nice to see happy Freeplane users :smile: 

@1. I had decided to preserve root node's core text and mindmap's file name so that there is some way to identify the mindmap after it is obfuscated (and shared). If User needs the information obfuscated, this can be done by hand, like you pointed out. 
Alternatively, this could be made into an option in preferences (e.g. obfuscate root text, obfuscate file name) – I'll think about it.

@2. The action `Insert->Image->Add image…` creates a link to the image file and Freeplane reads the image data from the disk each time the mindmap is opened for viewing. It would be quite hard to share a map with such external images, because it would require sharing the whole filesystem structure. Similarly, it would make little sense to insert a test image instead. Therefore the add-on obfuscates external-image paths to avoid revealing potentially sensitive information, e.g. "c:\User\**JohnDoe**\Documents\my_image.png"

@3. A good catch! Fixed. It will be released with a new version.

Regards
Mac

**Comment by autoca** [comment]
@2:
Yes, the logic behind it is quite obvious but, IMHO, flawed.
It would not require sharing the whole filesystem structure if the path is made relative. Not sure if this is possible via the UI, but with a text editor it is quite possible. If the image resides in the same folder as the mindmap, then it could be defined inside the mm file like that:
`<hook URI="file:test.png" SIZE="0.29069766" NAME="ExternalObject"/>`
I did not try other relative paths, but the same folder should be good enough, since it reveals nothing.

Besides, in general, IMHO it would make a lot of sense to replace all images with a default replacement image. I am sure you remember the bug I just reported recently. Without an image in the map it could not be recreated/tested. Also, it visually helps with orientation if there are images and not "error" messages in the nodes.
Maybe the user could specify in some setting which image should be the replacement. It should be enough if all existing images are replaced by the same image.

But I acknowledge that this might be some work to implement. The mere existance of the add-on as it is now is already awesome. So, don't stress yourself out over it.

Thanks again, regards, a.

EDIT:
Ok, the part about the relative path was still true in Freeplane-Setup-1.10.5-pre04. Seems like in Freeplane-Setup-1.10.5-pre05 it does not work anymore. I am testing more atm.

EDIT2:
Ok, in Freeplane-Setup-1.10.5-pre05 it must look like this:
`<hook URI="test.png" SIZE="0.29069766" NAME="ExternalObject"/>`

**Comment by euu2021** [comment]
To add another idea: sometimes, the name of the styles can be sensitive, and the user can easily forget to check it before sharing. So, i t would be nice to change it, too.

**Comment by autoca** [comment]
> sometimes, the name of the styles can be sensitive, and the user can easily forget to check it before sharing. So, i t would be nice to change it, too.

agreed! good thought!

**Comment by macmarrum** [comment]
AFAIK, Freeplane saves the path to an (external) image as absolute and there is no way in API to change it to relative. It might be a good Feature Request: for FP to respect `Preferences…->Environment->Hyperlink types->Links` (Relative, Absolute) when inserting external images.

Freeplane lacks the functionality to rename styles. I'll think about how to work around it.

**Comment by autoca** [comment]
Both can be done with search  and replace in the mm file (with a text editor outside of FP), or am I wrong?

**Comment by macmarrum** [comment]
Yes. That's the workaround I've had in mind.

**Comment by macmarrum** [comment]
Thanks for all your suggestions. Here's a new version: [MapObfuscator-2022.10.17.addon.mm](https://github.com/macmarrum/freeplane-MapObfuscator/releases/download/v2022.10.17/MapObfuscator-2022.10.17.addon.mm) 

Added preferences
Added style names obfuscation

**Comment by autoca** [comment]
This is amazing. Thank you for your friendly contribution!
If you ever want to improve it even further, I found at least 2 more ideas:
1) labels inside connectors (beginning, middle, and end)
2) When an image path is obfuscated then it seems that it is based on a relative path, which is good, but it also seems that the first sub-folder of that relative path is not getting obfuscated. (Maybe it is because my relative path started with 2 underscores: "__"?)

Don't get me wrong, I am really grateful. Really helpful work, as is. Thank you!
My thoughts are only meant as friendly feedback in case you are bored one day.

Regards, a.

**Comment by macmarrum** [comment]
@autoca Many thanks for your feedback. It's much appreciated.
1. Connector labels are already obfuscated. If it does not work, please create a [new issue](https://github.com/macmarrum/freeplane-MapObfuscator/issues/new)
2. Image paths should get obfuscated, also the first sub-folder. Please create a [new issue](https://github.com/macmarrum/freeplane-MapObfuscator/issues/new) and I'll look into it.

**Comment by autoca** [comment]
Ok, I will try to create these issues eventually.

One other thing that I forgot to mention:
The add-on also seems to only obfuscate the core-ascii chars. At least, ä ö ü ß (German keyboard) are not obfuscated.
This is not a real problem since it does not reveal any actual info, I think, other than the map author used such chars.
But I just wanted to mention it.

**Comment by macmarrum** [comment]
> The add-on also seems to only obfuscate the core-ascii chars. At least, ä ö ü ß (German keyboard) are not obfuscated.

A good point, thank you. I'll extend the relevant regex to cover also non-ASCII letters.

**Comment by macmarrum** [comment]
Non-ASCII characters are covered in [v2022.10.22](https://github.com/macmarrum/freeplane-MapObfuscator/releases)
